### PR TITLE
🦥 Slowmode Command 

### PR DESCRIPTION
**Pull request by Ikinon**
# Pull request

 - [x] I have read and agreed to the [code of conduct](https://github.com/CascadeBot/CascadeBot/blob/master/.github/CODE_OF_CONDUCT.md).
 - [x] I have read and complied with the [contributing guidelines](https://github.com/CascadeBot/CascadeBot/blob/master/.github/CONTRIBUTING.md).
 - [x] What I'm implementing was assigned to me or was not being worked on and is a somewhat requested feature. For reference see our [GitHub projects](https://github.com/orgs/CascadeBot/projects).
 - [x] I have tested all of my changes.
 
## Added/Changed
Added slowmode command

## Feature
Was requested in issues

### CODE REVIEW

**Comment by binaryoverload** [comment]
I've not had the chance to do a full review just yet but so far this looks pretty close to spot on. Good work!

Full review coming soon 😊

File: src/main/kotlin/org/cascadebot/cascadebot/commands/moderation/SlowmodeCommand.kt [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        if (!context.selfMember.hasPermission(channel, Permission.MANAGE_CHANNEL)) {
+            context.uiMessaging.sendBotDiscordPermError(Permission.MANAGE_CHANNEL)
+            return
+        }
+        val duration = ParserUtils.parseTextTime(context.getArg(0), false) // In milliseconds
+        if (duration / 1000 > 21600) {  // 6 hour maximum time imposed by API
+            context.typedMessaging.replyWarning(context.i18n("commands.slowmode.time_exceeded"))
+            return
+        }
+        if (duration.toInt() == 0) {
```
Comment (binaryoverload) [inline src/main/kotlin/org/cascadebot/cascadebot/commands/moderation/SlowmodeCommand.kt:40]:
```suggestion
        if (duration == 0L) {
```

File: src/main/kotlin/org/cascadebot/cascadebot/commands/moderation/SlowmodeResetSubCommand.kt [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+                return
+            }
+        }
+        if (!context.selfMember.hasPermission(channel, Permission.MANAGE_CHANNEL)) {
+            context.uiMessaging.sendBotDiscordPermError(Permission.MANAGE_CHANNEL)
+            return
+        }
+
+        channel.manager.setSlowmode(0).queue() // Slowmode off
+        context.typedMessaging.replySuccess(context.i18n("commands.slowmode.reset.reset_success", channel.name))
```
Comment (binaryoverload) [inline src/main/kotlin/org/cascadebot/cascadebot/commands/moderation/SlowmodeResetSubCommand.kt:29]: Use the success and failure consumers on the `queue` method like: 

```kotlin
queue({ do success }, {throwable -> do failure})
```

File: src/main/kotlin/org/cascadebot/cascadebot/commands/moderation/SlowmodeCommand.kt [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+            context.typedMessaging.replyDanger(context.i18n("responses.invalid_duration"))
+            return
+        }
+        channel.manager.setSlowmode(duration.toInt() / 1000).queue()
+        val interval: String = FormatUtils.formatTime(duration, context.locale, true).replace("(0[hms])".toRegex(), "")
+        context.typedMessaging.replySuccess(context.i18n("commands.slowmode.success", interval, channel.name))
```
Comment (binaryoverload) [inline src/main/kotlin/org/cascadebot/cascadebot/commands/moderation/SlowmodeCommand.kt:46]: Please also use the success and failure consumers for this one too. Make sure to have an error condition

File: src/main/kotlin/org/cascadebot/cascadebot/commands/moderation/SlowmodeResetSubCommand.kt (hunk 2 above)
Comment (binaryoverload) [reply src/main/kotlin/org/cascadebot/cascadebot/commands/moderation/SlowmodeResetSubCommand.kt:29]: Make sure to also have an error condition

File: src/main/resources/lang/en-GB.json [hunk 4]
```diff
@@ -100,6 +100,17 @@
       "guild_member_count": "Member Count",
       "guild_id": "ID: {0}"
     },
+    "slowmode": {
+      "command": "slowmode",
+      "description": "Sets the slowmode of a channel",
+      "time_exceeded": "Maximum time for slowmode of channels is 6 hours",
+      "invalid_channel" : "I cannot find channel `{0}`",
```
Comment (binaryoverload) [inline src/main/resources/lang/en-GB.json:107]: This should already be under responses.

File: src/main/resources/lang/en-GB.json [hunk 5]
```diff
@@ -100,6 +100,17 @@
...
+      "success": "Slowmode set to {0} in {1}",
+      "reset" : {
+        "description": "Resets the slowmode to off",
```
Comment (binaryoverload) [inline src/main/resources/lang/en-GB.json:110]: Please also include "command" for this

File: src/main/resources/lang/en-GB.json (hunk 4 above)
Comment (binaryoverload) [reply src/main/resources/lang/en-GB.json:107]: https://github.com/CascadeBot/CascadeBot/pull/239/files#diff-f7944c7ae4f73ab92f3d2297e07d5aa9R1261

File: src/main/kotlin/org/cascadebot/cascadebot/commands/moderation/SlowmodeCommand.kt [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    override fun onCommand(sender: Member, context: CommandContext) {
+        if (context.args.isEmpty()) {
+            context.uiMessaging.replyUsage()
+            return
+        }
+        var channel: TextChannel = context.channel
+        if (context.args.size == 2) {
+            try {
+                channel = DiscordUtils.getTextChannel(context.guild, context.getArg(1))
+            } catch (e: IllegalStateException) {
```
Comment (binaryoverload) [inline src/main/kotlin/org/cascadebot/cascadebot/commands/moderation/SlowmodeCommand.kt:26]: Need to look into this, IllegalStateException shouldn't be thrown

**Review by binaryoverload — CHANGES_REQUESTED** [review]

File: src/main/resources/arguments.json [hunk 7]
```diff
@@ -53,6 +53,19 @@
...
   },
+  "slowmode*": {
+    "interval*": {
+      "_type": "required",
+      "channel*": {
+        "_type": "optional"
+      }
+    },
+    "reset*" : {
+      "channel*": {
```
Comment (binaryoverload) [inline src/main/resources/arguments.json:64]:
```suggestion
      "channel": {
```

File: src/main/resources/arguments.json [hunk 8]
```diff
@@ -53,6 +53,19 @@
       "_type": "optional"
     }
   },
+  "slowmode*": {
+    "interval*": {
+      "_type": "required",
+      "channel*": {
```
Comment (binaryoverload) [inline src/main/resources/arguments.json:59]:
```suggestion
      "channel": {
```

**Review by binaryoverload — CHANGES_REQUESTED** [review]

**Review by binaryoverload — APPROVED** [review]

File: src/main/kotlin/org/cascadebot/cascadebot/scheduler/ActionType.kt [hunk 9]
```diff
@@ -1,12 +1,31 @@
...
-    REMINDER(ScheduledAction.ReminderActionData::class, { action -> TODO() });
+    REMINDER(ScheduledAction.ReminderActionData::class, { action -> TODO() }),
+
+    UNSLOWMODE(ScheduledAction.SlowmodeActionData::class, { action ->
+        if (action.data is ScheduledAction.SlowmodeActionData) {
+            action.guild?.let { guild ->
+                val targetChannel = guild.getGuildChannelById(action.data.targetId)
+                targetChannel?.manager?.setSlowmode(action.data.oldSlowmode)?.queue(null, {
+                    if (it is PermissionException) {
+                        action.channel?.let { channel -> Messaging.sendMessage(MessageType.DANGER, channel, i18n(action.guildId, "responses.no_discord_perm_bot", it.permission)) }
```
Comment (binaryoverload) [inline src/main/kotlin/org/cascadebot/cascadebot/scheduler/ActionType.kt:21]:
```suggestion
                        action.channel?.let { channel -> Messaging.sendMessage(MessageType.DANGER, channel, i18n(action.guildId, "responses.no_discord_perm_bot", it.permission.name)) }
```

**Review by binaryoverload — APPROVED** [review]
Just a small change please but otherwise great! :)

File: src/main/kotlin/org/cascadebot/cascadebot/commands/moderation/TempSlowmodeCommand.kt [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+
+        var channel: TextChannel = context.channel
+        if (context.args.size == 3) {
+            val tempChannel = DiscordUtils.getTextChannel(context.guild, context.getArg(2))
+            if (tempChannel != null) {
+                channel = tempChannel
+            } else {
+                context.typedMessaging.replyDanger(context.i18n("responses.cannot_find_channel_matching", context.getArg(1)))
+                return
+            }
```
Comment (weeryan17) [inline src/main/kotlin/org/cascadebot/cascadebot/commands/moderation/TempSlowmodeCommand.kt:50]: Why are you creating a temp channel variable here? You can just store it in the channel object, and if channel is null return. temp channel is a extra variable that doesn't need to exist.

File: src/main/kotlin/org/cascadebot/cascadebot/scheduler/ScheduledAction.kt [hunk 11]
```diff
@@ -22,6 +26,12 @@ data class ScheduledAction(
     val delay: Long
         get() = ChronoUnit.MILLIS.between(Instant.now(), executionTime).coerceAtLeast(0L)
 
+    val guild: Guild?
+        get() = CascadeBot.INS.shardManager.getGuildById(guildId)
+
+    val channel: TextChannel?
+        get() = CascadeBot.INS.shardManager.getTextChannelById(channelId)
+
```
Comment (weeryan17) [inline src/main/kotlin/org/cascadebot/cascadebot/scheduler/ScheduledAction.kt:34]: Are you up to date with dev? cause I'm fairly sure these exist on dev.

File: src/main/kotlin/org/cascadebot/cascadebot/commands/moderation/TempSlowmodeCommand.kt (hunk 10 above)
Comment (binaryoverload) [reply src/main/kotlin/org/cascadebot/cascadebot/commands/moderation/TempSlowmodeCommand.kt:50]: That could actually work. This was they way I suggested, didn't think of that Ryan 👌

Comment (Ikinon) [reply src/main/kotlin/org/cascadebot/cascadebot/commands/moderation/TempSlowmodeCommand.kt]: I may be testing it wrong but it seems as if channel cannot be null, so binary's solution probably is the way to go about this

File: src/main/kotlin/org/cascadebot/cascadebot/commands/moderation/TempSlowmodeCommand.kt [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+            return
+        }
+
+        val longDuration = ParserUtils.parseTextTime(context.getArg(1), false)
+        if (longDuration <= 0) {
+            context.typedMessaging.replyDanger(context.i18n("responses.invalid_duration"))
+            return
+        }
+
+        var channel: TextChannel = context.channel
```
Comment (weeryan17) [inline src/main/kotlin/org/cascadebot/cascadebot/commands/moderation/TempSlowmodeCommand.kt:42]: You can change this to `TextChannel?` to allow nulls.

Comment (binaryoverload) [reply src/main/kotlin/org/cascadebot/cascadebot/commands/moderation/TempSlowmodeCommand.kt:42]: > You can change this to `TextChannel?` to allow nulls.

Not advisable as it affects later code

Comment (weeryan17) [reply src/main/kotlin/org/cascadebot/cascadebot/commands/moderation/TempSlowmodeCommand.kt:42]: Ok then maybe he needs to have the temp channel then

Comment (weeryan17) [reply src/main/kotlin/org/cascadebot/cascadebot/commands/moderation/TempSlowmodeCommand.kt:42]: I don't like it as its an extra variable that could not exist, but megh.

**Pull request merged**